From 437df41cae8a4daa556e18a9225a7891ab1f3361 Mon Sep 17 00:00:00 2001
From: "peter.lin" <peter.lin@amlogic.com>
Date: Fri, 10 Jan 2014 14:25:51 +0800
Subject: [PATCH 3079/5965] PD #85986: avoid null pointer in
 wdev_cleanup_work(),add mutex clean_mtx for wdev_cleanup_work

---
 include/net/cfg80211.h | 1 +
 net/wireless/core.c    | 4 ++++
 net/wireless/nl80211.c | 1 +
 net/wireless/sme.c     | 4 ++++
 4 files changed, 10 insertions(+)
 mode change 100644 => 100755 include/net/cfg80211.h
 mode change 100644 => 100755 net/wireless/core.c
 mode change 100644 => 100755 net/wireless/nl80211.c
 mode change 100644 => 100755 net/wireless/sme.c

diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
old mode 100644
new mode 100755
index 26b5b692c22b..da6e2c542971
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -2857,6 +2857,7 @@ struct wireless_dev {
 	spinlock_t mgmt_registrations_lock;
 
 	struct mutex mtx;
+	struct mutex clean_mtx;
 
 	struct work_struct cleanup_work;
 
diff --git a/net/wireless/core.c b/net/wireless/core.c
old mode 100644
new mode 100755
index 64fcbae020d2..e0cbfaefb884
--- a/net/wireless/core.c
+++ b/net/wireless/core.c
@@ -803,6 +803,7 @@ static void wdev_cleanup_work(struct work_struct *work)
 	wake_up(&rdev->dev_wait);
 
 	dev_put(wdev->netdev);
+	mutex_unlock(&wdev->clean_mtx);
 }
 
 void cfg80211_unregister_wdev(struct wireless_dev *wdev)
@@ -913,6 +914,7 @@ static int cfg80211_netdev_notifier_call(struct notifier_block *nb,
 		 * are added with nl80211.
 		 */
 		mutex_init(&wdev->mtx);
+		mutex_init(&wdev->clean_mtx);
 		INIT_WORK(&wdev->cleanup_work, wdev_cleanup_work);
 		INIT_LIST_HEAD(&wdev->event_list);
 		spin_lock_init(&wdev->event_lock);
@@ -957,6 +959,7 @@ static int cfg80211_netdev_notifier_call(struct notifier_block *nb,
 		cfg80211_leave(rdev, wdev);
 		break;
 	case NETDEV_DOWN:
+		mutex_lock(&wdev->clean_mtx);
 		cfg80211_update_iface_num(rdev, wdev->iftype, -1);
 		dev_hold(dev);
 		queue_work(cfg80211_wq, &wdev->cleanup_work);
@@ -972,6 +975,7 @@ static int cfg80211_netdev_notifier_call(struct notifier_block *nb,
 			mutex_lock(&rdev->devlist_mtx);
 			rdev->opencount--;
 			mutex_unlock(&rdev->devlist_mtx);
+			mutex_unlock(&wdev->clean_mtx);
 			dev_put(dev);
 		}
 		cfg80211_update_iface_num(rdev, wdev->iftype, 1);
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
old mode 100644
new mode 100755
index 448c034184e2..9a01c1282c14
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -2483,6 +2483,7 @@ static int nl80211_new_interface(struct sk_buff *skb, struct genl_info *info)
 		 * through the netdev notifier and must be added here
 		 */
 		mutex_init(&wdev->mtx);
+		mutex_init(&wdev->clean_mtx);
 		INIT_LIST_HEAD(&wdev->event_list);
 		spin_lock_init(&wdev->event_lock);
 		INIT_LIST_HEAD(&wdev->mgmt_registrations);
diff --git a/net/wireless/sme.c b/net/wireless/sme.c
old mode 100644
new mode 100755
index e2f74e66a169..1afbf0be8904
--- a/net/wireless/sme.c
+++ b/net/wireless/sme.c
@@ -329,6 +329,10 @@ static void __cfg80211_sme_scan_done(struct net_device *dev)
 void cfg80211_sme_scan_done(struct net_device *dev)
 {
 	struct wireless_dev *wdev = dev->ieee80211_ptr;
+	if (wdev == NULL) {
+		printk("cfg80211_sme_scan_done : null pointer\n");
+		return;
+	}
 
 	wdev_lock(wdev);
 	__cfg80211_sme_scan_done(dev);
-- 
2.19.0

